File: src/main/java/net/arvandor/numinoustreasury/profession/NuminousCharacterProfessionRepository.java
package net.arvandor.numinoustreasury.profession;

import static net.arvandor.numinoustreasury.jooq.Tables.NUMINOUS_CHARACTER_PROFESSION;

import com.rpkit.characters.bukkit.character.RPKCharacterId;
import org.jooq.DSLContext;

import java.util.Objects;

public final class NuminousCharacterProfessionRepository {

    private final DSLContext dsl;

    public NuminousCharacterProfessionRepository(DSLContext dsl) {
        this.dsl = dsl;
    }

    public String getProfessionId(RPKCharacterId characterId) {
        return dsl.select(NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID)
                .from(NUMINOUS_CHARACTER_PROFESSION)
                .where(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.getValue()))
                .fetchOne(NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID);
    }

    public void setProfessionId(RPKCharacterId characterId, String professionId) {
        dsl.insertInto(NUMINOUS_CHARACTER_PROFESSION)
                .set(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID, characterId.getValue())
                .set(NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID, professionId)
                .set(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE, 0)
                .onConflict(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID).doUpdate()
                .set(NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID, professionId)
                .set(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE, 0)
                .where(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.getValue()))
                .execute();
    }

    public int getProfessionExperience(RPKCharacterId characterId) {
        return dsl.select(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
                .from(NUMINOUS_CHARACTER_PROFESSION)
                .where(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.getValue()))
                .fetchOptional(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
                .orElse(0);
    }

    public void getAndUpdate(RPKCharacterId characterId, ExperienceUpdateFunction function, ExperienceUpdateCallback callback) {
        dsl.transaction(config -> {
            DSLContext transactionalDsl = config.dsl();
            Integer experience = transactionalDsl
                    .select(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
                    .from(NUMINOUS_CHARACTER_PROFESSION)
                    .where(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.getValue()))
                    .forUpdate()
                    .fetchOne(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE);
            function.invoke(transactionalDsl, Objects.requireNonNullElse(experience, 0));
            Integer newExperience = transactionalDsl
                    .select(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
                    .from(NUMINOUS_CHARACTER_PROFESSION)
                    .where(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.getValue()))
                    .fetchOne(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE);
            callback.invoke(Objects.requireNonNullElse(experience, 0), Objects.requireNonNullElse(newExperience, 0));
        });
    }

    public void setProfessionExperience(RPKCharacterId characterId, int experience) {
        setProfessionExperience(dsl, characterId, experience);
    }

    public void setProfessionExperience(DSLContext dsl, RPKCharacterId characterId, int experience) {
        dsl.update(NUMINOUS_CHARACTER_PROFESSION)
                .set(NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE, experience)
                .where(NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.getValue()))
                .execute();
    }

}

File: src/main/kotlin/net/arvandor/numinoustreasury/profession/NuminousCharacterProfessionRepository.kt
package net.arvandor.numinoustreasury.profession

import com.rpkit.characters.bukkit.character.RPKCharacterId
import net.arvandor.numinoustreasury.jooq.Tables
import org.jooq.Configuration
import org.jooq.DSLContext

class NuminousCharacterProfessionRepository(private val dsl: DSLContext) {
    fun getProfessionId(characterId: RPKCharacterId): String? {
        return dsl.select(Tables.NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID)
            .from(Tables.NUMINOUS_CHARACTER_PROFESSION)
            .where(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.value))
            .fetchOne(Tables.NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID)
    }

    fun setProfessionId(
        characterId: RPKCharacterId,
        professionId: String?,
    ) {
        dsl.insertInto(Tables.NUMINOUS_CHARACTER_PROFESSION)
            .set(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID, characterId.value)
            .set(Tables.NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID, professionId)
            .set(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE, 0)
            .onConflict(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID).doUpdate()
            .set(Tables.NUMINOUS_CHARACTER_PROFESSION.PROFESSION_ID, professionId)
            .set(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE, 0)
            .where(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.value))
            .execute()
    }

    fun getProfessionExperience(characterId: RPKCharacterId): Int {
        return dsl.select(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
            .from(Tables.NUMINOUS_CHARACTER_PROFESSION)
            .where(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.value))
            .fetchOptional(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
            .orElse(0)
    }

    fun getAndUpdate(
        characterId: RPKCharacterId,
        function: ExperienceUpdateFunction,
        callback: ExperienceUpdateCallback,
    ) {
        dsl.transaction { config: Configuration ->
            val transactionalDsl = config.dsl()
            val experience =
                transactionalDsl
                    .select(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
                    .from(Tables.NUMINOUS_CHARACTER_PROFESSION)
                    .where(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.value))
                    .forUpdate()
                    .fetchOne(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
            function.invoke(transactionalDsl, experience ?: 0)
            val newExperience =
                transactionalDsl
                    .select(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
                    .from(Tables.NUMINOUS_CHARACTER_PROFESSION)
                    .where(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.value))
                    .fetchOne(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE)
            callback.invoke(experience ?: 0, newExperience ?: 0)
        }
    }

    fun setProfessionExperience(
        characterId: RPKCharacterId,
        experience: Int,
    ) {
        setProfessionExperience(dsl, characterId, experience)
    }

    fun setProfessionExperience(
        dsl: DSLContext,
        characterId: RPKCharacterId,
        experience: Int,
    ) {
        dsl.update(Tables.NUMINOUS_CHARACTER_PROFESSION)
            .set(Tables.NUMINOUS_CHARACTER_PROFESSION.EXPERIENCE, experience)
            .where(Tables.NUMINOUS_CHARACTER_PROFESSION.CHARACTER_ID.eq(characterId.value))
            .execute()
    }
}
